demos/iconvert.1
.\"
.\" $Id: iconvert.1,v 1.1 2003/04/06 15:52:38 leeming Exp $
.\"
.\" Use the following command to generate viewable man page
.\"   tbl fdesign.1 | {nt}roff -man
.\"
.\"   ul | more -fsk to view
.\"
.\"
.TH ICONVERT 1L "March 1999" "Version 0.11" "Forms Library"
.SH NAME
iconvert \-  Forms Library Image Support demo
.\"
.\" setup
.de Cr
.ie n (c)
.el \(co
..
.SH SYNOPSIS
.if n iconvert
.if t  \{
.B iconvert
\}
[-options] input output [fmt]
.SH DESCRIPTION
.I iconvert
is a demo program based on 
.BR  "Forms Library" 's
image support. The program can be used to convert image formats.
The input image file, if is in one of the supported format, is automatically 
recognized and read, then the output file is written into a format
specified by the extension of the output file name 
or by the parameter fmt if present. 
.SH OPTIONS
.I iconvert
accepts the following command line options
.TP 6
.B \-help
Prints a brief summary of all command line options. 
.TP
.B \-version
Prints the current version of
.I iconvert
and exits.
.TP
.B \-verbose
Specifies verbose mode where each phase of the conversion is printed
to stderr.

.SH EXIT STATUS
The command exits with status 0 if the conversion is successful;
1 if the command line is bad, and 3 if conversion failed.
.SH SEE ALSO
xforms(5), fdesign(1L)
.SH AUTHOR
Copyright
.Cr
1999 T.C. Zhao (tc_zhao@yahoo.com)
.SH HOME PAGE
http://bragg.phys.uwm.edu/xforms


NEWS
------------------------------------------
V1.1 ???????????
  o Reset the horizontal offset of a browser to zero if the
    horizontal scrollbar is turned off (Bug #3205.)
  o Add a fl_replace_xyplot_point_in_overlay function that generalizes
    the existing fl_replace_xyplot_point.
  o Add new accessor functions to query and modify the timeouts that
    control the behaviour of the slider and counter widgets.
  o Enable input widgets to (partially) display characters larger than
    the input height.
  o Pass the associated (XEvent * xev) to fl_handle_object on an FL_DRAW
    event, allowing an FL_FREE object to make use of this info to
    redraw only the part of the window that has changed.
  o Fix compilation on SCO OpenServer, AIX and systems without GNU make.
------------------------------------------
V1.0.91pre5 April xx, 2008
  o Memory allocation bug in fl_initialize() fixed.
  o Output of fdesign should now be valid C (ANSI 89)
  o Number of symbols and length of names not limited anymore.
  o Public functions added to set or query for buttons to which
    mouse buttons they react:
     fl_set_button_mouse_buttons()
     fl_get_button_mouse_buttons()
    Per default buttons again react to all mouse buttons.
  o Removed restriction that required one of a set of radio buttons to
    be always active active.
  o Changes to fdesign to set this behaviour.
------------------------------------------
V1.0.91pre4 March 31, 2008
  o Just some bug fixes
------------------------------------------
V1.0.91pre3 March 26, 2008
  o If input object has focus and another object gets clicked on
    the input object loses the focus, resulting in an evaluation
    of the content of the input object.
  o Timeouts got a higher priority to make them more precise.
  o Some reorganization of memory allocation handling in lots
    of places to be able to really deallocate unused memory.
  o Buttons now only react to the left mouse button.
  o Enforcing that always one of a set of radio buttons is
    active.
------------------------------------------
V1.0.91pre2 March 20, 2008
  o Cleanup of popup code.
  o Scroll wheel can now be used for dials.
  o Scrollbars on textbrowsers etc. now have correct size and position.
  o FL_DROPLIST_CHOICE choice objects now work.
  o Scroll wheel support to walk through the items of a menu.
  o New public function:
      fl_get_choice_item_mode()
------------------------------------------
V1.0.91pre1 March 12, 2008
  o Resizing code rewritten, several redraw bugs removed.
  o Double clicking on a file in the file selector repaired.
  o Three new public functions:
      functions fl_get_object_bw()
	  fl_get_object_gravity()
      fl_get_object_resize()
  o Bugs fixed that resulted in attempts to draw on non-existent
    windows which then resulted in X errors.
  o Corrections to make sliders on text browsers work correctly
  o Number of forms that can be created is now not limited anymore.
  o Object and event queue code rewritten to allow queues that grow
    dynamically if necessary.
  o Fixed bug that sometimes resulted in fl_do_forms() return an
    object when the scroll wheel was used.
------------------------------------------
V1.0.90 March 30, 2004
  o Migrate from Imake to the GNU autotools.
  o "Composite" widgets, such as the browser, can now display tooltips.
  o Prevent a crash when rotating grayscale images by 90 degree multiples.
  o Enable xforms to handle XPixmaps containing colour "opaque".
  o Tabfolder widgets are now resizable.
  o Re-add fl_lookup_RGBcolor which fell off the dist at 1.0pre3.
  o fl_init_RGBdatabase is no longer needed, does nothing and is
    retained for compatibility only.
  o Pass KeyRelease events to the widgets.
  o Display the tab forms correctly when using bottom tab folders.
  o Enable xforms to read a wider variety of gif and jpeg images.
  o Enable fdesign to work correctly with paths like ../foo/bar/your_file.fd.
  o Add a -dir <destdir> option to fdesign to place generated files in destdir.
  o Add an RPM spec file. The XForms rpms can be built with a simple
    "make rpmdist".
  o Get rid of the separate snprintf compatibility library.
  o Removal of lots of little pieces of cruft that crept into the xforms
    1.0 release.
------------------------------------------
V1.0RC6.1 November 19, 2002
  o. Fixed early tooltip hide causing recursive fl_free_forms() entry.
  o. Fixed typo in `lib/Imakefile'.

------------------------------------------
V1.0RC6 November 16, 2002
  o. TabFolder coordinate update corrected. (Angus Leeming)
  o. Tooltips could be orphaned when form hidden.  (Angus Leeming)

------------------------------------------
V1.0RC5.2 October 2, 2002
  o. Documented `InstallNonstandard' procedure.

------------------------------------------
V1.0RC5.1 October 1, 2002
  o. Mac OS X build support.

------------------------------------------
V1.0RC3 June 5, 2002
  o. Fixed condition in forms.c preventing use of cursor keys.
  o. Fumigated with -Wall, eliminating all complaints in xforms
     sources.

------------------------------------------
V1.0RC2	June 1, 2002
  o. Picked up some missing prototypes in `flimage.h'.
  o. Deleted fl_snprintf() with snprintf() and changed all references
     thereto.  Added a new library for those lacking the latter
     function, filched from the SpamAssassin distributon, of all things.
     Had to hack it a bit to get it to work.  Others, on other systems,
     may have to as well.

------------------------------------------
V1.0RC1	May 30, 2002
  o. Very minor API changes.  `FL_ERROR_FUNC' has been typedef'd for
     fl_set_error_handler() function.  `FL_VAL_FILTER' has also been
     typedef'd for use by fl_set_{counter,slider}_filter().  Neither
     of these changes should affect much of anything.
  o. The flimage functions are in their own library now: libflimage.
     They should be usable standalone, but I have not had an
     opportunity to test them.

demos/iconvert.c
/*
 *  This file is part of XForms.
 *
 *  XForms is free software; you can redistribute it and/or modify it
 *  under the terms of the GNU Lesser General Public License as
 *  published by the Free Software Foundation; either version 2.1, or
 *  (at your option) any later version.
 *
 *  XForms is distributed in the hope that it will be useful, but
 *  WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with XForms; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 59 Temple Place - Suite 330, Boston,
 *  MA 02111-1307, USA.
 */


/*
 * Convert an image file using the image support of Forms Library.
 *
 *
 *  Usage: iconvert [-version][-verbose][-help] inputimage outimage [fmt]
 *     output image format is determined by the extension or
 *     by fmt if present.
 *
 *   Exit status:  0 (success) 1 (bad command line)  3 (conversion failed)
 *
 *  T.C. Zhao (03/1999)
 *
 */

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <stdlib.h>
#include "include/forms.h"
#include "flimage.h"

static void initialize( void );
static void usage( const char *,
				   int );
static int parse_command_line( int *,
							   char ** );

static char *version="$Id: iconvert.c,v 1.7 2008/12/27 22:20:31 jtt Exp $";


/***************************************
 ***************************************/

int
main( int    argc,
	  char * argv[ ] )
{
    FL_IMAGE *im;
    const char *fmt = 0;
    char *const *args;

    initialize( );
    args = argv + parse_command_line( &argc, argv );

    if (   argc < 3 ||
		 ! ( fmt = argc >= 4 ? args[ 3 ] : strrchr( args[ 2 ], '.' ) ) )
		usage( argv[ 0 ], argc >= 3 );

    fmt += fmt[ 0 ] == '.';
    im = flimage_load( strdup( args[ 1 ] ) );
    return flimage_dump( im, args[ 2 ], fmt ) < 0 ? 3 : 0;
}


/***************************************
 ***************************************/

static void
usage( const char * cmd,
	   int          more )
{
    const FLIMAGE_FORMAT_INFO *info;
    int n,
		i,
		k;

    fprintf( stderr, "Usage: %s [-verbose][-help] infile outfile [fmt]\n",
			 cmd );

    if ( ! more )
       exit( 1 );

    fputs( " The output format is determined by the file extension "
		   "or fmt.\n fmt or extension must be one of the following\n",
		   stderr );

    n = flimage_get_number_of_formats( );
    for ( i = 1, k = 0; i <= n; i++ )
    {
          info = flimage_get_format_info( i );
          if ( info->read_write & FLIMAGE_WRITABLE )
          {
			  fprintf( stderr, "\t%s", info->extension );
			  if ( ++k % 6 == 0 )
				  fputc( '\n', stderr );
          }
    }

    if ( k % 6 )
        fputc( '\n', stderr );
    exit( 1 );
}


/* shut up visual_cue */

/***************************************
 ***************************************/

static int
noop( FL_IMAGE   * im  FL_UNUSED_ARG,
	  const char * s   FL_UNUSED_ARG )
{
    return 0;
}


/***************************************
 ***************************************/

static int
parse_command_line( int  * argc,
					char * argv[ ] )
{
    int i;
    static FLIMAGE_SETUP setup;

    setup.visual_cue = noop;

    for ( i = 1; i < *argc && *argv[ i ] == '-'; )
    {
		if ( strncmp( argv[ i ], "-verb", 5 ) == 0 )
		{
			setup.visual_cue = 0;
			i++;
		}
		else if ( strncmp( argv[ i ], "-vers", 5 ) == 0 )
		{
            fprintf( stderr, "%s\n", version );
            exit( 0 );
         }
		else if ( strncmp( argv[ i ], "-h", 2 ) == 0 )
		{
			usage( argv[ 0 ], 1 );
			i++;
		}
		else
			usage( argv[ 0 ], 0 );
    }

    flimage_setup( &setup );
    *argc -= i - 1;
    return i - 1;
}


/***************************************
 ***************************************/

static void
initialize( void )
{
    flimage_enable_xpm( );
    flimage_enable_gif( );
    flimage_enable_bmp( );
    flimage_enable_sgi( );
    flimage_enable_fits( );
    flimage_enable_png( );
    flimage_enable_xwd( );
    flimage_enable_tiff( );
    flimage_enable_ps( );
#ifndef NO_JPEG
    flimage_enable_jpeg( );
#endif
}
